### PR TITLE
Add new lint `while_float`

### LINKED ISSUE #758: Lint float loops
those are usually better written as integer-y loops + cast within the loop

Note that this is mostly about `while` loops.

### PR DESCRIPTION
**Pull request by yusufraji**
This PR adds a nursery lint that checks for while loops comparing floating point values.

changelog:
```
changelog: [`while_float`]: Checks for while loops comparing floating point values.
```

Fixes #758

### CODE REVIEW

File: clippy_lints/src/loops/mod.rs [hunk 1]
```diff
@@ -416,6 +417,39 @@ declare_clippy_lint! {
...
+    /// ### What it does
+    /// Checks for while loops comparing floating point values.
+    ///
+    /// ### Why is this bad?
+    /// If you increment floating point values, errors can compound,
+    /// so, use integers instead if possible.
+    ///
+    /// ### Known problems
+    /// The lint will catch all while loops comparing floating point
+    /// values no matter whether it's a threshold or something.
```
Comment (llogiq) [inline clippy_lints/src/loops/mod.rs:430]:
```suggestion
    /// values without regarding the increment.
```

**Comment by llogiq** [comment]
Thank you for making clippy better! This is a good starting point. We might want to extend the lint to look into the loop variant and check if the increment is integer, but as this is a nursery lint, we might well do that in a follow-up PR.

So I just left a small doc suggestion, otherwise this looks good to merge.

**Comment by llogiq** [comment]
Ok, let's merge this and see how it fares.

@bors r+

If you'd like to keep working on this lint, here are some ideas for possible future directions:

1. Try to extract a constant from the comparison arguments and check if it is integral. If not, avoid linting.
2. Try to extract the loop variable and walk the loop body to find the increment(s), then also check that those are integral.
3. (once 2. is in place), visit the whole method looking for uses of that variable and check if all can be converted to integer math. If so, add a multi-suggestion changing all found uses. Note: This is pretty advanced stuff already.

**Comment by xFrednet** [comment]
Is this lint supposed to be in `nursery`? Seems a bit odd for a new lint :thinking: If so, we should have some issues for followup fixes

**Pull request merged**